### PR TITLE
Uplift `RegionVid`, `TermKind` to `rustc_type_ir`, and `EagerResolver` to `rustc_next_trait_solver`

### PR DESCRIPTION
**Pull request by compiler-errors**
- Uplift `RegionVid`. This was complicated due to the fact that we implement `polonius_engine::Atom` for `RegionVid` -- but I just separated that into `PoloniusRegionVid`, and added `From`/`Into` impls so it can be defined in `rustc_borrowck` separately. Coherence 😵 
- Change `InferCtxtLike` to expose `opportunistically_resolve_{ty,ct,lt,int,float}_var` so that we can uplift `EagerResolver` for use in the canonicalization methods.
- Uplift `TermKind` much like `GenericArgKind`

All of this is miscellaneous dependencies for making more `EvalCtxt` methods generic.

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? lcnr

File: compiler/rustc_type_ir/src/region_kind.rs [hunk 1]
```diff
@@ -1,13 +1,35 @@
...
-use rustc_macros::{TyDecodable, TyEncodable};
+use rustc_macros::{HashStable_NoContext, TyDecodable, TyEncodable};
 use std::fmt;
 
 use crate::{DebruijnIndex, DebugWithInfcx, InferCtxtLike, Interner, WithInfcx};
 
 use self::RegionKind::*;
 
+rustc_index::newtype_index! {
+    /// A **ty**pe **v**ariable **ID**.
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/region_kind.rs:12]: this comment is wrong :3

File: compiler/rustc_next_trait_solver/src/canonicalizer.rs [hunk 2]
```diff
@@ -388,13 +397,8 @@ impl<Infcx: InferCtxtLike<Interner = I>, I: Interner> TypeFolder<I>
...
                 ty::InferConst::Var(vid) => {
                     assert_eq!(
-                        self.infcx.root_ct_var(vid),
-                        vid,
-                        "region vid should have been resolved fully before canonicalization"
-                    );
-                    assert_eq!(
-                        self.infcx.probe_ct_var(vid),
-                        None,
+                        self.infcx.opportunistic_resolve_ct_var(vid, ty).kind(),
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/canonicalizer.rs:400]: pls add comment for why we have to compare `kind` here

File: compiler/rustc_next_trait_solver/src/canonicalizer.rs [hunk 3]
```diff
@@ -467,3 +471,82 @@ impl<I: Interner> TypeFolder<I> for RegionsToStatic<I> {
         }
     }
 }
+
+///////////////////////////////////////////////////////////////////////////
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/canonicalizer.rs:475]: can you move that into a separate module, e.g. `rustc_next_rait_solver/src/resolve.rs` or sth

**Comment by lcnr** [comment]
r=me after nits

**Pull request merged**